### PR TITLE
LibGfx+LibGL+LibGPU+LibSoftGPU: Implement blend equations

### PR DESCRIPTION
**Pull request by gmta**
Used by the SRB2 port (which still crashes when trying to compile shaders). :^)

Screenshot_20240218_154306

### CODE REVIEW

**Comment by gmta** [comment]
Made it a draft for now, want to check whether the golden test can include alpha transparency by setting alpha to 0 in `glClearColor`.

File: Tests/LibGL/TestRender.cpp [hunk 1]
```diff
@@ -317,3 +318,48 @@ TEST_CASE(0011_tex_env_combine_with_constant_color)
...
+        GL_MAX,
+    };
+    auto constexpr grid_size = blend_modes.size();
+    auto constexpr cell_size = 2.f / grid_size;
+    for (size_t x = 0; x < grid_size; ++x) {
+        for (size_t y = 0; y < grid_size; ++y) {
+            auto rgb_mode = blend_modes[x];
+            auto alpha_mode = blend_modes[y];
+
+            glBlendEquationSeparate(rgb_mode, alpha_mode);
```
Comment (nico) [inline Tests/LibGL/TestRender.cpp:351]: Want to add a glGet() test right after while here? (Up to you)

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 2]
```diff
@@ -173,6 +173,42 @@ ALWAYS_INLINE static Vector4<f32x4> get_blend_factor(GPU::BlendFactor blend_fact
...
+    };
+
+    // Single calculation if RGB mode is the same as the alpha mode
+    if (rgb_mode == alpha_mode)
+        return apply_equation(rgb_mode, source, source_weights, destination, destination_weights);
+
+    // Split RGB/alpha calculation
+    auto rgb = apply_equation(rgb_mode, source.xyz(), source_weights.xyz(), destination.xyz(), destination_weights.xyz());
+    auto alpha = apply_equation(alpha_mode, source.w(), source_weights.w(), destination.w(), destination_weights.w());
+    return { rgb.x(), rgb.y(), rgb.z(), alpha };
```
Comment (nico) [inline Userland/Libraries/LibSoftGPU/Device.cpp:209]: Oh, we don't use premul alpha?

**Review by nico — APPROVED** [review]
Looks good as far as I can tell!

(Pretty new to LibSoftGPU, but know OpenGL pretty well. I'm hoping to learn more about LibSoftGPU at some point!)

**Comment by nico** [comment]
(Red CI is real though.)

File: Tests/LibGL/TestRender.cpp (hunk 1 above)
Comment (gmta) [reply Tests/LibGL/TestRender.cpp:358]: Good idea, I've also added two assertions for the initial state.

File: Userland/Libraries/LibSoftGPU/Device.cpp (hunk 2 above)
Comment (gmta) [reply Userland/Libraries/LibSoftGPU/Device.cpp:209]: That's up to you, with `glBlendFunc(Separate)` you can choose how each component is transformed. `glBlendEquation(Separate)` only changes the operation performed on the source color/weights and destination color/weights.

**Comment by gmta** [comment]
> (Red CI is real though.)

Let's all pretend I did not forget to add the right golden test. :^)

**Pull request merged**